### ISSUE TITLE
Layouts within layouts won't render

### ISSUE THREAD

**Issue by samollari** [samollari/LaunchControl#1]
Trying to set up multiple ULXD racks
```js
new VerticalLayoutComponent([
    new ULXD4QComponent(...),
    new ULXD4QComponent(...),
])
```
fails to render

Looks like there's some problem with `partialRender` with layouts in layouts

**Comment by samollari** [comment]
Simplifying this:
```js
new VerticalLayoutComponent([
    new HorizontalLayoutComponent([ new BubbleWrapComponent(), new BubbleWrapComponent() ]),
    new HorizontalLayoutComponent([ new BubbleWrapComponent(), new BubbleWrapComponent() ])
])
```
Renders the first time, but touching a cell calls `partialRender`, failing with the same error:
```
Uncaught Error: Element is not a member of the given array!
    assertIsElementOf util.ts:104
    partialRenderRequestedChild layouts.ts:36
    partialRender layouts.ts:114
    renderComponent renderer.ts:35
    requestRender renderer.ts:72
    set requestRender/</component.requestRender layouts.ts:30
    set requestRender/</component.requestRender layouts.ts:30
    touched bubblewrap.ts:20
    touched layouts.ts:81
    touched layouts.ts:127
    renderInteractiveComponentToLaunchpad renderer.ts:66
    emit events.js:153
    init launchpad.ts:143
    init launchpad.ts:126
    withMidiAccess main.ts:86
    main main.ts:48
    async* main.ts:128
```

> [!note]
> This also occurs with other versions of nesting layouts (vertical within horizontal, horizontal within horizontal both tested)

**Comment by samollari** [comment]
After about an hour of meticulously mentally stepping through the logic flow, I think I found the issue.

`renderInteractiveComponentToLaunchpad` sets the root component's `requestRender` function:
```ts
component.requestRender = (componentTrace) => {
    const canvas = Canvas.renderComponent(
        component,
        componentTrace.slice(1),
    );
    renderCanvasToLaunchpad(canvas, position, launchpad);
};
```
[source](https://github.com/i-am-gizm0/LaunchControl/blob/95dade430eaef3c57e66a2871a016f1d0d4e1e48/src/layout/renderer.ts#L74)

Removing the first element of `componentTrace` with `.slice(1)` sets up a `Canvas.renderComponent` call asking the root component to directly render a child two levels below it. The observed error is thrown when the root component tries to `.partialRenderRequestedChild` with the component trace consisting of the child two levels below it. `assertIsElementOf` correctly flags this grandchild as not a direct child, and aborts rendering.

Simply removing the `.slice(1)` from setting up the root component's `requestRender` should resolve this issue: the root component asks its direct child to partial render as it should.

[My notes thinking through the problem](https://gist.github.com/i-am-gizm0/bf429784977b7501aefd587bbe5aa07c)
[Change made in this commit](https://github.com/i-am-gizm0/LaunchControl/commit/63e5ea970806e46af288526506ee437be8470163)

**Comment by samollari** [comment]
Fixing that uncovered another bug in the `partialRender` methods of the `Vertical` and `HorizontalLayoutComponent`s: they would always try to place their updated child after the rest of their content (they summed the relevant dimension of *all* their children when they should have only summed up to (not including) the updating child)

[Change made in this commit](https://github.com/i-am-gizm0/LaunchControl/commit/2d704428539dca4c7f6fe56f148b60747b4a3a39)

**Comment by samollari** [comment]
Fixing that uncovered another bug in `renderCanvasToLaunchpad`: `undefined` colors in a `Canvas` were getting passed to create a `StandardLEDColorDefinition`, which was correctly flagging this as an invalid value and aborting displaying.

[Change made in this commit](https://github.com/i-am-gizm0/LaunchControl/commit/49f1bada500446e955b8503fc47a3d273aa56949)

**Comment by samollari** [comment]
3fe875d fixes a type bug introduced in 49f1bad
